docs/src/design/proposals/Registration.md

```diff
@@ -0,0 +1,518 @@
+   arguments, disambiguated with JSDoc. The software does not take part
+   in dependency management among plugins; rather, this responsibility is
+   plainly communicated to developers.
+2. Extension points are made explicit; `Provider` instances must be
+   reachable for plugins to configure, and may be made available as
+   public fields of `Plugin`s. Their types can be clearly documented,
+   usages and interactions can be followed with standard developer tools
+   (e.g. breakpoints), and so on.
+3. Reuse of classes between plugins is neither facilitated nor impeded
+   by the registration API. If, however, plugins are written following the
+   "expose classes in namespaces" approach, then it is trivially to
+   expose additional classes in these same namespaces.
+
+This solution offers further benefits:
+
+* The `Provider` API is robust enough to support the various existing
+  extensions of Open MCT, but its usage is opt-in; plugins are free
+  to expose other (potentially wildly different) means of extension.
+  Usage of `Provider`s is _encouraged_ to promote ubiquitous
+  extensibility, but no limitation is exposed.
+* Plugins (and even whole applications, e.g. `MCT`) can be unit tested,
+  in the same style as individual extensions.
+* By moving everything into classes which accept dependencies, a
+  degree of inflexibility is removed from the architecture. In principle,
+  it should be possible to run multiple instances of `MCT` (with
+  their own service instances, etc.) within the same environment. While
+  this is not specifically desirable, it reflects a generally looser
+  coupling between the software and it environment (no expectation of a
+  `bundles.json`, no usage of global state at the language level or
+  effectively-global state at the RequireJS level, etc.) and implies
+  greater flexibility of the application's components.
+* Provides a reduction in code associated with a particular capability;
+  the Framework Layer consists of 18 classes (and has broader
+  responsibilities, including script-loading) whereas this approach
+  consists of 6 classes. (Although some additional work for Angular
+  integration and legacy support will also be needed before achieving
+  parity with the Framework Layer.)
+
+There are some problems with this approach:
+
+* It is highly sensitive to ordering; does not address the problem of
+  [separating configuration from use](http://www.martinfowler.com/articles/injection.html#SeparatingConfigurationFromUse),
+  but instead leaves this as a problem to solve with code style
+  (requiring familiarity with the system.) This is particularly
+  true with `Provider#get` (don't want to invoke before configuration
+  is finished), but also true for `Application` and `Plugin`.
+  * One approach to mitigate this would be to throw `Error`s when
+    calls are made out-of-order (e.g. configuration after use.)
+* Some redundancy among interfaces (`Plugin` and `Application` both
```

Comment:
I do like the idea of moving extension points off of plugins and onto some other thing. Sticking this on `Application` instances seems reasonable, and doesn't impede other applications from introducing more extension points. That said, I'm also feeling that these extension points just need to live on another object (perhaps `Application<T>` delivers a `T` to its plugins to configure?) and would like to have a way for these groupings to be a bit more modular (why does my plugin need access to telemetry-related extensions if it's registering a persistence provider? it would be nice if we didn't force plugin developers to violate the Law of Demeter.) It's not obvious to me how to reconcile this, but I'd like to look for a reasonable solution.